examples/psbt.rs

```diff
@@ -0,0 +1,711 @@
+}
+
+fn main() {
+    let secp = SECP256K1;
+
+    //
+    // Step 0: Create the extended private key from the test vector data.
+    //
+
+    let ext_priv = build_extended_private_key();
+    let ext_pub = ExtendedPubKey::from_priv(&secp, &ext_priv);
+    let fingerprint = ext_pub.fingerprint();
+
+    //
+    // Step 1: The creator.
+    //
+
+    let tx = create_transaction();
+    let psbt = create_psbt(tx);
+
+    //
+    // Step 2: The first updater.
+    //
+
+    // QUESTION: This update implies the updater knows the extended public key,
+    // is that correct? Otherwise where does the fingerprint come from?
+    let psbt = update_psbt(psbt, fingerprint);
+
+    //
+    // Step 3: The second updater.
+    //
+
+    let psbt = update_psbt_with_sighash_all(psbt);
+
+    //
+    // Step 4: The first signer.
+    //
+
+    // Strings from BIP174 test vector.
+    let test_vector = vec![
+        ("cP53pDbR5WtAD8dYAW9hhTjuvvTVaEiQBdrz9XPrgLBeRFiyCbQr", "m/0h/0h/0h"),
+        ("cR6SXDoyfQrcp4piaiHE97Rsgta9mNhGTen9XeonVgwsh4iSgw6d", "m/0h/0h/2h"),
+    ];
+
+    // We pass the keys to the signer after doing verification to make explicit
+    // that signer is only using these two keys.
+    let keys = parse_and_verify_keys(&ext_priv, &test_vector);
+    let psbt_1 = signer_one_sign(psbt.clone(), &keys);
+
+    //
+    // Step 5: The second signer.
+    //
+
+    // Strings from BIP174 test vector.
+    let test_vector = vec![
+        ("cT7J9YpCwY3AVRFSjN6ukeEeWY6mhpbJPxRaDaP5QTdygQRxP9Au", "m/0h/0h/1h"),
+        ("cNBc3SWUip9PPm1GjRoLEJT6T41iNzCYtD7qro84FMnM5zEqeJsE", "m/0h/0h/3h"),
+    ];
+
+    let keys = parse_and_verify_keys(&ext_priv, &test_vector);
+    let psbt_2 = signer_two_sign(psbt, &keys);
+
+    //
+    // Step 6: Combiner the two signed PSBTs.
+    //
+
+    let combined = combine(psbt_1, psbt_2);
+
+    //
+    // Step 7: Finalize the PSBT.
+    //
+
+    let finalized = finalize(combined);
+
+    //
+    // Step 8: Extract the transaction.
+    //
+
+    let _tx = extract_transaction(finalized);
+
+    //
+    // Step 9: Test lexicographical PSBT combiner
+    //
+
+    combine_lexicographically();
+}
+
+/// Attempts to build an extended private key from seed and also directly from a string.
+fn build_extended_private_key() -> ExtendedPrivKey {
+    // Strings from BIP174 test vector.
+    let extended_private_key = "tprv8ZgxMBicQKsPd9TeAdPADNnSyH9SSUUbTVeFszDE23Ki6TBB5nCefAdHkK8Fm3qMQR6sHwA56zqRmKmxnHk37JkiFzvncDqoKmPWubu7hDF";
+    let seed = "cUkG8i1RFfWGWy5ziR11zJ5V4U4W3viSFCfyJmZnvQaUsd1xuF3T";
+
+    let ext_priv = ExtendedPrivKey::from_str(extended_private_key).unwrap();
+
+    // TODO: Fix this so parsing seed works.
+    let seed = base58::from(&seed).unwrap();
+    let seeded = ExtendedPrivKey::new_master(NETWORK, &seed).unwrap();
+    if seeded != ext_priv {
+        eprintln!("Keys from seed and tprv string do not match");
+    }
```

Comment:
Parsing the seed has me stumped too. When I picked this up `seed` was being parsed from base64. That, nor base58, nor base58check result in a matching `ext_priv`. Is this example used in any other project or are we gonna have to ask achow how he did this? I checked the NBitcoin implementation where they list the seed as `masterseed` in a json file of test data but it is never consumed.

Reply:
Yeah I tried base58 too :) I asked on bitcoin-dev mailing list.

Reply:
Got an answer from Andrew Chow, its formatted as a WIF private key, I have no idea why (I asked him that :).  The following code works
```
    // Strings from BIP174 test vector.
    let extended_private_key = "tprv8ZgxMBicQKsPd9TeAdPADNnSyH9SSUUbTVeFszDE23Ki6TBB5nCefAdHkK8Fm3qMQR6sHwA56zqRmKmxnHk37JkiFzvncDqoKmPWubu7hDF";
    let seed = "cUkG8i1RFfWGWy5ziR11zJ5V4U4W3viSFCfyJmZnvQaUsd1xuF3T";

    let xpriv = ExtendedPrivKey::from_str(extended_private_key).unwrap();

    let sk = PrivateKey::from_wif(&seed).unwrap();
    let seeded = ExtendedPrivKey::new_master(NETWORK, &sk.inner.secret_bytes()).unwrap();
    assert_eq!(xpriv, seeded);
```